test/query-path/run.sh

```diff
@@ -17,9 +17,27 @@ test -z "$(run_bob query-path --release root)"
 test -z "$(run_bob query-path --release root/interm1/child)"
 test -z "$(run_bob query-path --release root/interm2/child)"
 
+# Still, must report error on stderr as long as we don't provide -q
+test -n "$(run_bob query-path --release root 2>&1 | grep 'not present')"
+test -n "$(run_bob query-path --release root/interm1/child 2>&1 | grep 'not present')"
+test -n "$(run_bob query-path --release root/interm2/child 2>&1 | grep 'not present')"
+test -z "$(run_bob query-path --release -q root 2>&1 | grep 'not present')"
+test -z "$(run_bob query-path --release -q root/interm1/child 2>&1 | grep 'not present')"
+test -z "$(run_bob query-path --release -q root/interm2/child 2>&1 | grep 'not present')"
+
+# And must fail if --fail is provided
+test !$(run_bob query-path --release --fail root)
+test !$(run_bob query-path --release --fail root/interm1/child)
+test !$(run_bob query-path --release --fail root/interm2/child)
```

Comment:
I'm not sure if this is really testing what is should. I think the `!` is subject to history expansion. :thinking: Have you tested the test? :wink: 

I would suggest `expect_fail run_bob query-path --release --fail root`...

Reply:
Well, maybe I'm misreading [man test](https://linux.die.net/man/1/test), or maybe I'm even referring to the wrong man page, or - and that's the most probable case - my novice shell knowledge isn't sufficient, but

> ! EXPRESSION
>
>     EXPRESSION is false 

:thinking: May you enlighten me?

Yes, I've run the tests, but maybe I've misinterpreted their behaviour... :flushed:

I'll change it to your suggestion as soon as I get to it. :male_shrug:

Reply:
The `test` command gets the "expression" as argument. As such it only looks at strings that are passed to it. Apart from the different operators an empty string is "false" and everything else is "true". The exit status of `test` is 0 if the evaluated expression is true otherwise 1.

You wanted to test is the exit status of `run_bob`. What you instead tested is the output of `run_bob`. See this example:

```sh
#!/bin/bash -e

if test !$(/bin/false) ; then
   echo false is true
fi
if test !$(/bin/true) ; then
   echo true is true
fi
```

Output:

```
false is true
true is true
```

Typing the above example on the command line is another story and won't work for a different reason. The "!" is the history expansion meta character. See `HISTORY EXPANSION` in the bash man page for the gory details...

Reply:
Bash, nuff said...

But thanks for your comprehensive explanation. As I said, I'll change it once I get to it.